src/app/CASESessionManager.cpp

```diff
@@ -36,16 +36,53 @@ void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Cal
                                                 uint8_t attemptCount, Callback::Callback<OnDeviceConnectionRetry> * onRetry
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
 )
+{
+    FindOrEstablishSessionHelper(peerId, onConnection, onFailure, nullptr
+#if CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+                                 ,
+                                 attemptCount, onRetry
+#endif
+    );
+}
+
+void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                                Callback::Callback<OperationalSessionSetup::OnSetupFailure> * onSetupFailure
+#if CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+                                                ,
+                                                uint8_t attemptCount, Callback::Callback<OnDeviceConnectionRetry> * onRetry
+#endif
+)
+{
+    FindOrEstablishSessionHelper(peerId, onConnection, nullptr, onSetupFailure
+#if CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+                                 ,
+                                 attemptCount, onRetry
+#endif
+    );
+}
+
+void CASESessionManager::FindOrEstablishSessionHelper(const ScopedNodeId & peerId,
+                                                      Callback::Callback<OnDeviceConnected> * onConnection,
+                                                      Callback::Callback<OnDeviceConnectionFailure> * onFailure,
+                                                      Callback::Callback<OperationalSessionSetup::OnSetupFailure> * onSetupFailure
+#if CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+                                                      ,
+                                                      uint8_t attemptCount, Callback::Callback<OnDeviceConnectionRetry> * onRetry
+#endif
+)
 {
     ChipLogDetail(CASESessionManager, "FindOrEstablishSession: PeerId = [%d:" ChipLogFormatX64 "]", peerId.GetFabricIndex(),
                   ChipLogValueX64(peerId.GetNodeId()));
 
+    // Ensure that exactly one of onFailure or onSetupFailure is non-null
```

Comment:
This is an API change. The current API allows the failure callback to be null, no?

I just realized this also means we might need an overload that takes nullptr_t, because otherwise code that passes null will break.

Reply:
If you pass nullptr for onSetupFailure, the compiler will select the overload that matches the provided argument types.
Both overloaded function matches, but the final result will same after reaching the helper function since neither failure callback will be triggered.

Added  another overloaded func which takes nullptr_t to make it explicit